### PR TITLE
Make tcp connections timeout after 1sec

### LINKED ISSUE #7: Timeout doesn't work
In Client.cs we have this [line](https://github.com/nblockchain/TcpEchoSharp/blob/master/src/Client/Client.cs#L32):

```
socket.ReceiveTimeout = 500;
```

However, it's not working. We should fix this so that when a server it's not responding it doesn't take ages until it fails (in my machine it takes about 75 seconds to throw an exception).

The best way to describe this issue is by showing a test that doesn't work at the moment:

```
        [Test]
        public async Task ProperNonEternalTimeout()
        {
            var someRandomIP = "52.1.57.181";
            bool? succesful = null;

            var stopWatch = new Stopwatch();
            stopWatch.Start();
            try
            {
                var client = new TcpEcho.StratumClient(someRandomIP, 50001);
                var result = await client.BlockchainTransactionGet(
                    17,
                    "2f309ef555110ab4e9c920faa2d43e64f195aa027e80ec28e1d243bd8929a2fc"
                );
                succesful = true;
            }
            catch
            {
                succesful = false;
                stopWatch.Stop();
            }
            Assert.That(succesful.HasValue, Is.EqualTo(true), "test is broken?");
            Assert.That(succesful.Value, Is.EqualTo(false), "IP is not too random? port was open actually!");
            Assert.That(stopWatch.Elapsed, Is.LessThan(TimeSpan.FromSeconds(1)));
        }
```

I've tried to fix it with this patch:

https://gist.github.com/knocte/d3cca16a43b4131d9aff266e40ae3186

But it seems it's not enough, the code must be blocking somewhere else...

### PR DESCRIPTION
**Pull request by canndrew**
Fixes https://github.com/nblockchain/TcpEchoSharp/issues/7

### CODE REVIEW

File: src/Client/Client.cs [hunk 1]
```diff
@@ -31,7 +31,9 @@ public abstract class Client {
             using (Socket socket = new Socket (SocketType.Stream, ProtocolType.Tcp)) {
                 socket.ReceiveTimeout = 500;
 
-                await socket.ConnectAsync (endpoint, port);
+                if (!socket.ConnectAsync (endpoint, port).Wait (1000)) {
```
Comment (knocte) [inline src/Client/Client.cs]: let's change the .ReceiveTimeout to be 1000 too, and we store it in a constant

**Comment by canndrew** [comment]
For the test, it would make sense to connect to an address we know won't respond. Is there some way in C# we can open a TCP listening socket but not send back a SYN-ACK when our socket tries to connect? The only way I can think of is (on linux) to use network namespaces and tun/tap, since raw sockets require root/admin privileges.

**Comment by canndrew** [comment]
Ping @c410-f3r

**Comment by knocte** [comment]
Mmm, I think that using `.Wait` is a bad practice because it makes the thread block, thus ripping the advantages of using the async/await model. What we need is do something like this: https://stackoverflow.com/questions/4238345/asynchronously-wait-for-taskt-to-complete-with-timeout

**Comment by canndrew** [comment]
Right, of course. Fixed.

**Comment by knocte** [comment]
@canndrew this could have broken the already existing tests, so I added a new catch.

And I was going to merge the PR after this, however, I was curious to do a little experiment: remove the `break;` from the `ConnectWithElectrumServersTransactionGet` test, and the only effect should be that it takes a bit longer until it executes on all servers right? however, this PR seems to make it hang. Can you investigate? Putting some Console.WriteLine statements might help to easily find where it gets stuck.

**Comment by knocte** [comment]
s/all tests/all servers/

**Comment by knocte** [comment]
Nevermind, found the problem!

**Pull request merged**